### DISCUSSION TITLE
How to check if a transaction returned a rollback error?

### DISCUSSION THREAD

**Discussion by keepsimple1** [diesel-rs/diesel#2801] [Q&A] [answered]
https://github.com/diesel-rs/diesel/blob/3e8fd4d7ce257830f1735e9a31c9a5f0fb0d777f/diesel/src/connection/mod.rs#L52-L54

Referring to the above doc comment, how do we check if a transaction hit the second case, i.e. returned an error from the rollback?

**Comment by weiznich** [comment, marked as answer]
There is currently no way to actually check in a general way if the rollback was successful or not, as this could return nearly every error from a connection. I would be open for improvements here.

**Reply by keepsimple1** [reply]
How about adding a new variant in enum `diesel::result::Error` just for rollbacks?   and then use it in `transaction` method? Something like in this diff: 

```diff
diff --git a/diesel/src/connection/mod.rs b/diesel/src/connection/mod.rs
index c01dd4f74..f3d5af385 100644
--- a/diesel/src/connection/mod.rs
+++ b/diesel/src/connection/mod.rs
@@ -112,7 +112,7 @@ pub trait Connection: SimpleConnection + Sized + Send {
                 Ok(value)
             }
             Err(e) => {
-                Self::TransactionManager::rollback_transaction(self)?;
+                Self::TransactionManager::rollback_transaction(self).map_err(|e| Error::RollbackError(Box::new(e)))?;
                 Err(e)
             }
         }
diff --git a/diesel/src/result.rs b/diesel/src/result.rs
index 9b44aa171..f4746731f 100644
--- a/diesel/src/result.rs
+++ b/diesel/src/result.rs
@@ -62,6 +62,10 @@ pub enum Error {
     /// by PostgreSQL.
     SerializationError(Box<dyn StdError + Send + Sync>),

+    /// An error occurred during the rollback.
+    /// (or use `Box<dyn StdError + Send + Sync>`, which seems to compile faster.)
+    RollbackError(Box<Error>),
+
     /// Roll back the current transaction.
     ///
     /// You can return this variant inside of a transaction when you want to
@@ -283,6 +287,7 @@ impl Display for Error {
             Error::QueryBuilderError(ref e) => e.fmt(f),
             Error::DeserializationError(ref e) => e.fmt(f),
             Error::SerializationError(ref e) => e.fmt(f),
+            Error::RollbackError(ref e) => e.fmt(f),
             Error::RollbackTransaction => write!(f, "The current transaction was aborted"),
             Error::AlreadyInTransaction => write!(
                 f,
```

Then a caller can check if `transaction` failed with `Error::RollbackError` ? and use `.source()` to get to the cause?

**Reply by weiznich** [reply]
That would be an solution, but I would likely wrap a `diesel::result::Error` as inner type of `RollbackError`. 
Mind opening a PR for this? That should include a documentation change for `Connection::transaction` mentioning that this error type is returned for rollback errors.

**Reply by keepsimple1** [reply]
> That would be an solution, but I would likely wrap a `diesel::result::Error` as inner type of `RollbackError`.

Did you mean not to use `Box`?  Currently I'm wrapping a `Box<diesel::result::Error>` as inner type of `RollbackError`.  Rust seems not to allow me to wrap without `Box`.

> Mind opening a PR for this? That should include a documentation change for `Connection::transaction` mentioning that this error type is returned for rollback errors.

Will do.

**Reply by weiznich** [reply]
> Did you mean not to use Box? Currently I'm wrapping a Box<diesel::result::Error> as inner type of RollbackError. Rust seems not to allow me to wrap without Box.

That error message is correct, `Box` is required here. I did misread the type as `Box<dyn std::error:Error>`, so that's just a mistake on my side. Sorry for the confusion.

**Reply by keepsimple1** [reply]
Thanks for confirming. No problem :)

**Reply by keepsimple1** [reply]
Followed up in issue #2810 and PR #2811. Thanks for merging the PR.

**Reply by weiznich** [reply]
Thanks for writing that PR. Such contributions are always welcome :tada:
